File: sdks/rust/src/transport/http.rs
//! HTTP fallback transport.
//!
//! Wraps `reqwest::Client` behind the same [`Transport`] interface
//! the RPC path uses, so `NexusClient` can pick a transport once at
//! construction and stop branching on scheme for every method call.
//!
//! The translation from wire-level command names to HTTP routes is a
//! thin hard-coded table — every HTTP route the legacy CLI used has a
//! mapping here. Commands without a mapping surface as a structured
//! error so callers see exactly why the fallback path refused the
//! request.

use crate::error::{NexusError, Result};
use async_trait::async_trait;
use base64::Engine;
use nexus_protocol::rpc::types::NexusValue;
use reqwest::{Client, Method};
use serde_json::Value;

use super::endpoint::{Endpoint, Scheme};
use super::{Transport, TransportRequest, TransportResponse};

/// HTTP credentials — API key or basic auth. Both may be set; the key
/// takes precedence to match the CLI's behaviour.
#[derive(Debug, Clone, Default)]
pub struct HttpCredentials {
    pub api_key: Option<String>,
    pub username: Option<String>,
    pub password: Option<String>,
}

/// HTTP transport over an existing `reqwest::Client`. Uses the
/// endpoint's `as_http_url()` method so `nexus://` URLs still get a
/// reachable HTTP URL (swapping to the sibling port 15474).
pub struct HttpTransport {
    endpoint: Endpoint,
    client: Client,
    base_url: String,
    credentials: HttpCredentials,
}

impl HttpTransport {
    pub fn new(
        endpoint: Endpoint,
        credentials: HttpCredentials,
        timeout_secs: u64,
    ) -> Result<Self> {
        let base_url = endpoint.as_http_url();
        let client = Client::builder()
            .timeout(std::time::Duration::from_secs(timeout_secs))
            .user_agent(format!("nexus-sdk/{}", env!("CARGO_PKG_VERSION")))
            .build()
            .map_err(|e| NexusError::Configuration(format!("reqwest build: {e}")))?;
        Ok(Self {
            endpoint,
            client,
            base_url,
            credentials,
        })
    }

    fn auth(&self, mut req: reqwest::RequestBuilder) -> reqwest::RequestBuilder {
        if let Some(key) = &self.credentials.api_key {
            req = req.header("X-API-Key", key);
        } else if let (Some(u), Some(p)) = (&self.credentials.username, &self.credentials.password)
        {
            let token = base64::engine::general_purpose::STANDARD.encode(format!("{u}:{p}"));
            req = req.header("Authorization", format!("Basic {token}"));
        }
        req
    }

    async fn dispatch(&self, cmd: &str, args: &[NexusValue]) -> Result<Value> {
        match cmd {
            "CYPHER" => {
                let query = first_str(args).ok_or_else(|| arg_err("CYPHER", 0, "string"))?;
                let params = args.get(1).map(nexus_to_json).unwrap_or(Value::Null);
                let body = serde_json::json!({
                    "query": query,
                    "parameters": if params.is_null() { Value::Null } else { params },
                });
                let url = format!("{}/cypher", self.base_url);
                let resp = self
                    .auth(self.client.request(Method::POST, &url).json(&body))
                    .send()
                    .await
                    .map_err(NexusError::Http)?;
                http_json(resp).await
            }
            "PING" | "HEALTH" => {
                let url = format!("{}/health", self.base_url);
                let resp = self
                    .auth(self.client.request(Method::GET, &url))
                    .send()
                    .await
                    .map_err(NexusError::Http)?;
                http_json(resp).await
            }
            "STATS" => {
                let url = format!("{}/stats", self.base_url);
                let resp = self
                    .auth(self.client.request(Method::GET, &url))
                    .send()
                    .await
                    .map_err(NexusError::Http)?;
                http_json(resp).await
            }
            "EXPORT" => {
                let fmt = first_str(args).ok_or_else(|| arg_err("EXPORT", 0, "string"))?;
                let url = format!("{}/export?format={}", self.base_url, fmt);
                let resp = self
                    .auth(self.client.request(Method::GET, &url))
                    .send()
                    .await
                    .map_err(NexusError::Http)?;
                let text = resp.text().await.map_err(NexusError::Http)?;
                Ok(serde_json::json!({"format": fmt, "data": text}))
            }
            "IMPORT" => {
                let fmt = first_str(args).ok_or_else(|| arg_err("IMPORT", 0, "string"))?;
                let payload = args
                    .get(1)
                    .and_then(|v| v.as_str().map(String::from))
                    .ok_or_else(|| arg_err("IMPORT", 1, "string"))?;
                let url = format!("{}/import?format={}", self.base_url, fmt);
                let resp = self
                    .auth(self.client.request(Method::POST, &url).body(payload))
                    .send()
                    .await
                    .map_err(NexusError::Http)?;
                http_json(resp).await
            }
            other => Err(NexusError::Configuration(format!(
                "HTTP fallback does not know how to route '{other}' \
                 — add an entry to nexus-sdk/src/transport/http.rs::dispatch"
            ))),
        }
    }
}

#[async_trait]
impl Transport for HttpTransport {
    async fn execute(&self, req: TransportRequest) -> Result<TransportResponse> {
        let json = self.dispatch(&req.command, &req.args).await?;
        Ok(TransportResponse {
            value: json_to_nexus(json),
        })
    }

    fn describe(&self) -> String {
        match self.endpoint.scheme {
            Scheme::Https => format!("{} (HTTPS)", self.endpoint),
            _ => format!("{} (HTTP)", self.endpoint),
        }
    }

    fn is_rpc(&self) -> bool {
        false
    }
}

// ── Helpers ────────────────────────────────────────────────────────────────

fn first_str(args: &[NexusValue]) -> Option<String> {
    args.first().and_then(|v| v.as_str().map(String::from))
}

fn arg_err(cmd: &str, idx: usize, ty: &str) -> NexusError {
    NexusError::Configuration(format!(
        "HTTP fallback: '{cmd}' argument {idx} must be a {ty}"
    ))
}

async fn http_json(resp: reqwest::Response) -> Result<Value> {
    let status = resp.status();
    if status.is_success() {
        resp.json::<Value>().await.map_err(NexusError::Http)
    } else {
        let body = resp.text().await.unwrap_or_default();
        Err(NexusError::Api {
            message: body,
            status: status.as_u16(),
        })
    }
}

/// JSON → NexusValue (mirror of the CLI helper of the same name).
pub(crate) fn json_to_nexus(v: Value) -> NexusValue {
    match v {
        Value::Null => NexusValue::Null,
        Value::Bool(b) => NexusValue::Bool(b),
        Value::Number(n) => {
            if let Some(i) = n.as_i64() {
                NexusValue::Int(i)
            } else if let Some(f) = n.as_f64() {
                NexusValue::Float(f)
            } else {
                NexusValue::Float(n.as_u64().unwrap_or(0) as f64)
            }
        }
        Value::String(s) => NexusValue::Str(s),
        Value::Array(a) => NexusValue::Array(a.into_iter().map(json_to_nexus).collect()),
        Value::Object(m) => NexusValue::Map(
            m.into_iter()
                .map(|(k, v)| (NexusValue::Str(k), json_to_nexus(v)))
                .collect(),
        ),
    }
}

/// NexusValue → JSON (mirror of the CLI helper).
pub(crate) fn nexus_to_json(v: &NexusValue) -> Value {
    match v {
        NexusValue::Null => Value::Null,
        NexusValue::Bool(b) => Value::Bool(*b),
        NexusValue::Int(i) => Value::Number((*i).into()),
        NexusValue::Float(f) => serde_json::Number::from_f64(*f)
            .map(Value::Number)
            .unwrap_or(Value::Null),
        NexusValue::Str(s) => Value::String(s.clone()),
        NexusValue::Bytes(b) => Value::Array(b.iter().copied().map(Value::from).collect()),
        NexusValue::Array(a) => Value::Array(a.iter().map(nexus_to_json).collect()),
        NexusValue::Map(pairs) => {
            let mut obj = serde_json::Map::new();
            for (k, val) in pairs {
                let key = match k {
                    NexusValue::Str(s) => s.clone(),
                    NexusValue::Int(i) => i.to_string(),
                    other => format!("{:?}", other),
                };
                obj.insert(key, nexus_to_json(val));
            }
            Value::Object(obj)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn json_roundtrip_covers_primitive_variants() {
        let cases = [
            Value::Null,
            Value::Bool(true),
            Value::from(42i64),
            Value::from(3.25f64),
            Value::from("hi"),
        ];
        for case in cases {
            let back = nexus_to_json(&json_to_nexus(case.clone()));
            assert_eq!(back, case);
        }
    }

    #[test]
    fn json_roundtrip_handles_nested_map_and_array() {
        let src = serde_json::json!({
            "labels": ["Person"],
            "properties": {"name": "Alice", "age": 30}
        });
        let back = nexus_to_json(&json_to_nexus(src.clone()));
        assert_eq!(back, src);
    }

    #[tokio::test]
    async fn http_fallback_rejects_unknown_command() {
        let ep = Endpoint::parse("http://127.0.0.1:1").unwrap();
        let t = HttpTransport::new(ep, HttpCredentials::default(), 5).unwrap();
        let err = t
            .dispatch("WIDGET", &[])
            .await
            .expect_err("unknown cmd must error");
        assert!(format!("{err}").contains("does not know how to route"));
    }
}

File: sdks/go/transport/http.go
package transport

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"time"
)

// HttpTransport wraps a net/http client with the same [Transport]
// interface the RPC path uses. A thin hard-coded route table maps
// wire-level verbs onto the REST endpoints the legacy Go client
// relied on. Unknown verbs surface [ErrUnmappedCommand].
type HttpTransport struct {
	endpoint Endpoint
	creds    Credentials
	baseURL  string
	client   *http.Client
}

// NewHttpTransport builds a fresh HTTP transport.
func NewHttpTransport(endpoint Endpoint, creds Credentials, timeout time.Duration) *HttpTransport {
	if timeout == 0 {
		timeout = 30 * time.Second
	}
	return &HttpTransport{
		endpoint: endpoint,
		creds:    creds,
		baseURL:  endpoint.AsHttpURL(),
		client:   &http.Client{Timeout: timeout},
	}
}

// Execute implements [Transport].
func (t *HttpTransport) Execute(ctx context.Context, req Request) (Response, error) {
	val, err := t.dispatch(ctx, req.Command, req.Args)
	if err != nil {
		return Response{}, err
	}
	return Response{Value: val}, nil
}

// Describe implements [Transport].
func (t *HttpTransport) Describe() string {
	tag := "HTTP"
	if t.endpoint.Scheme == "https" {
		tag = "HTTPS"
	}
	return fmt.Sprintf("%s (%s)", t.endpoint, tag)
}

// IsRpc implements [Transport].
func (t *HttpTransport) IsRpc() bool { return false }

// Close implements [Transport]. net/http has no persistent socket to free.
func (t *HttpTransport) Close() error {
	t.client.CloseIdleConnections()
	return nil
}

func (t *HttpTransport) applyAuth(req *http.Request) {
	if t.creds.APIKey != "" {
		req.Header.Set("X-API-Key", t.creds.APIKey)
	} else if t.creds.Username != "" && t.creds.Password != "" {
		token := base64.StdEncoding.EncodeToString([]byte(t.creds.Username + ":" + t.creds.Password))
		req.Header.Set("Authorization", "Basic "+token)
	}
}

func (t *HttpTransport) dispatch(ctx context.Context, cmd string, args []NexusValue) (NexusValue, error) {
	switch cmd {
	case "CYPHER":
		query, ok := args[0].AsString()
		if !ok {
			return NexusValue{}, fmt.Errorf("HTTP fallback: 'CYPHER' argument 0 must be a string")
		}
		body := map[string]any{"query": query}
		if len(args) > 1 {
			body["parameters"] = NexusToJson(args[1])
		} else {
			body["parameters"] = nil
		}
		return t.doJSON(ctx, http.MethodPost, "/cypher", body)
	case "PING", "HEALTH":
		return t.doJSON(ctx, http.MethodGet, "/health", nil)
	case "STATS":
		return t.doJSON(ctx, http.MethodGet, "/stats", nil)
	case "DB_LIST":
		return t.doJSON(ctx, http.MethodGet, "/databases", nil)
	case "DB_CREATE":
		name, ok := args[0].AsString()
		if !ok {
			return NexusValue{}, fmt.Errorf("HTTP fallback: 'DB_CREATE' argument 0 must be a string")
		}
		return t.doJSON(ctx, http.MethodPost, "/databases", map[string]any{"name": name})
	case "DB_DROP":
		name, ok := args[0].AsString()
		if !ok {
			return NexusValue{}, fmt.Errorf("HTTP fallback: 'DB_DROP' argument 0 must be a string")
		}
		return t.doJSON(ctx, http.MethodDelete, "/databases/"+url.PathEscape(name), nil)
	case "DB_USE":
		name, ok := args[0].AsString()
		if !ok {
			return NexusValue{}, fmt.Errorf("HTTP fallback: 'DB_USE' argument 0 must be a string")
		}
		return t.doJSON(ctx, http.MethodPut, "/session/database", map[string]any{"name": name})
	case "DB_CURRENT":
		return t.doJSON(ctx, http.MethodGet, "/session/database", nil)
	case "LABELS":
		return t.doJSON(ctx, http.MethodGet, "/schema/labels", nil)
	case "REL_TYPES":
		return t.doJSON(ctx, http.MethodGet, "/schema/relationship-types", nil)
	case "EXPORT":
		fmtStr, ok := args[0].AsString()
		if !ok {
			return NexusValue{}, fmt.Errorf("HTTP fallback: 'EXPORT' argument 0 must be a string")
		}
		text, err := t.doText(ctx, http.MethodGet, "/export?format="+url.QueryEscape(fmtStr), nil, "")
		if err != nil {
			return NexusValue{}, err
		}
		return JsonToNexus(map[string]any{"format": fmtStr, "data": text}), nil
	case "IMPORT":
		fmtStr, ok := args[0].AsString()
		if !ok {
			return NexusValue{}, fmt.Errorf("HTTP fallback: 'IMPORT' argument 0 must be a string")
		}
		payload, ok := args[1].AsString()
		if !ok {
			return NexusValue{}, fmt.Errorf("HTTP fallback: 'IMPORT' argument 1 must be a string")
		}
		return t.doRaw(ctx, http.MethodPost, "/import?format="+url.QueryEscape(fmtStr), payload, "text/plain")
	}
	return NexusValue{}, &ErrUnmappedCommand{Command: cmd}
}

func (t *HttpTransport) doJSON(ctx context.Context, method, path string, body any) (NexusValue, error) {
	var reqBody io.Reader
	if body != nil {
		data, err := json.Marshal(body)
		if err != nil {
			return NexusValue{}, err
		}
		reqBody = bytes.NewReader(data)
	}
	req, err := http.NewRequestWithContext(ctx, method, t.baseURL+path, reqBody)
	if err != nil {
		return NexusValue{}, err
	}
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	t.applyAuth(req)
	resp, err := t.client.Do(req)
	if err != nil {
		return NexusValue{}, err
	}
	defer resp.Body.Close()
	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return NexusValue{}, err
	}
	if resp.StatusCode >= 400 {
		return NexusValue{}, &HttpError{StatusCode: resp.StatusCode, Body: string(raw)}
	}
	if len(raw) == 0 {
		return NxNull(), nil
	}
	var decoded any
	if err := json.Unmarshal(raw, &decoded); err != nil {
		// Fall back to string if JSON decode fails.
		return NxStr(string(raw)), nil
	}
	return JsonToNexus(decoded), nil
}

func (t *HttpTransport) doText(ctx context.Context, method, path string, body io.Reader, contentType string) (string, error) {
	req, err := http.NewRequestWithContext(ctx, method, t.baseURL+path, body)
	if err != nil {
		return "", err
	}
	if contentType != "" {
		req.Header.Set("Content-Type", contentType)
	}
	t.applyAuth(req)
	resp, err := t.client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}
	if resp.StatusCode >= 400 {
		return "", &HttpError{StatusCode: resp.StatusCode, Body: string(raw)}
	}
	return string(raw), nil
}

func (t *HttpTransport) doRaw(ctx context.Context, method, path, body, contentType string) (NexusValue, error) {
	req, err := http.NewRequestWithContext(ctx, method, t.baseURL+path, bytes.NewBufferString(body))
	if err != nil {
		return NexusValue{}, err
	}
	req.Header.Set("Content-Type", contentType)
	t.applyAuth(req)
	resp, err := t.client.Do(req)
	if err != nil {
		return NexusValue{}, err
	}
	defer resp.Body.Close()
	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return NexusValue{}, err
	}
	if resp.StatusCode >= 400 {
		return NexusValue{}, &HttpError{StatusCode: resp.StatusCode, Body: string(raw)}
	}
	if len(raw) == 0 {
		return NxNull(), nil
	}
	var decoded any
	if err := json.Unmarshal(raw, &decoded); err != nil {
		return NxStr(string(raw)), nil
	}
	return JsonToNexus(decoded), nil
}
